From 7e5a943afdda7aa7ec22f65e19d17d7ab1386ead Mon Sep 17 00:00:00 2001
From: Dong Aisheng <b29396@freescale.com>
Date: Thu, 9 Apr 2015 16:16:09 +0800
Subject: [PATCH 0542/5242] MLK-10629-2 mmc: sdhci-esdhc-imx: implement
 wifi_card_detect function

commit  b007250914a6da3eb35a888dfbb5f9304c164d86 from
https://source.codeaurora.org/external/imx/linux-imx.git

WiFi driver could call wifi_card_detect function to re-detect card,
this is required by some special WiFi cards like broadcom WiFi.
To use this function, a new property is introduced to indicate a wifi host.

Signed-off-by: Dong Aisheng <b29396@freescale.com>
(cherry picked from commit 74e71dd0aebb9e931f02aefa3dd1990cbe642ae4)
Signed-off-by: Haibo Chen <haibo.chen@freescale.com>

Conflicts:
	Documentation/devicetree/bindings/mmc/fsl-imx-esdhc.txt
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 .../devicetree/bindings/mmc/fsl-imx-esdhc.txt      |    2 ++
 drivers/mmc/host/sdhci-esdhc-imx.c                 |   13 +++++++++++++
 2 files changed, 15 insertions(+)

diff --git a/Documentation/devicetree/bindings/mmc/fsl-imx-esdhc.txt b/Documentation/devicetree/bindings/mmc/fsl-imx-esdhc.txt
index 3e29050..04667b9 100644
--- a/Documentation/devicetree/bindings/mmc/fsl-imx-esdhc.txt
+++ b/Documentation/devicetree/bindings/mmc/fsl-imx-esdhc.txt
@@ -35,6 +35,8 @@ Optional properties:
   This property allows user to change the tuning step to more than one delay
   cells which is useful for some special boards or cards when the default
   tuning step can't find the proper delay window within limited tuning retries.
+- wifi-host : assigned as a wifi host.
+  This is required for some WiFi cards to do card detect
 
 Examples:
 
diff --git a/drivers/mmc/host/sdhci-esdhc-imx.c b/drivers/mmc/host/sdhci-esdhc-imx.c
index a9960f9..c05bd17 100644
--- a/drivers/mmc/host/sdhci-esdhc-imx.c
+++ b/drivers/mmc/host/sdhci-esdhc-imx.c
@@ -147,6 +147,14 @@
 /* A clock frequency higher than this rate requires strobe dll control */
 #define ESDHC_STROBE_DLL_CLK_FREQ	100000000
 
+static struct mmc_host *wifi_mmc_host;
+void wifi_card_detect(void)
+{
+	WARN_ON(!wifi_mmc_host);
+	mmc_detect_change(wifi_mmc_host, 0);
+}
+EXPORT_SYMBOL(wifi_card_detect);
+
 struct esdhc_soc_data {
 	u32 flags;
 };
@@ -1185,6 +1193,11 @@ static void sdhci_esdhc_imx_hwinit(struct sdhci_host *host)
 	if (mmc_gpio_get_cd(host->mmc) >= 0)
 		host->quirks &= ~SDHCI_QUIRK_BROKEN_CARD_DETECTION;
 
+	if (of_get_property(np, "wifi-host", NULL)) {
+		wifi_mmc_host = host->mmc;
+		dev_info(mmc_dev(host->mmc), "assigned as wifi host\n");
+	}
+
 	return 0;
 }
 #else
-- 
1.7.9.5

